File: scripts/test_pattern/test_adapter.py
"""Tests for the Adapter pattern."""

import sys
import os
import math
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from adapter import RoundHole, RoundPeg, SquarePeg, SquarePegAdapter


class TestAdapter(unittest.TestCase):

    EPS = 1e-6

    # ------------------------------------------------------------------
    # RoundHole / RoundPeg
    # ------------------------------------------------------------------
    def test_round_peg_fits_larger_hole(self):
        self.assertTrue(RoundHole(5.0).fits(RoundPeg(4.0).radius()))

    def test_round_peg_does_not_fit_smaller_hole(self):
        self.assertFalse(RoundHole(3.0).fits(RoundPeg(4.0).radius()))

    def test_round_peg_fits_exactly(self):
        self.assertTrue(RoundHole(5.0).fits(RoundPeg(5.0).radius()))

    # ------------------------------------------------------------------
    # Adapter radius calculation
    # ------------------------------------------------------------------
    def test_adapter_radius_is_half_diagonal(self):
        adapter = SquarePegAdapter(SquarePeg(2.0))
        expected = 2.0 * math.sqrt(2) / 2
        self.assertAlmostEqual(expected, adapter.radius(), places=6)

    def test_small_square_peg_fits_in_big_hole(self):
        hole    = RoundHole(5.0)
        adapter = SquarePegAdapter(SquarePeg(4.0))
        self.assertTrue(hole.fits(adapter.radius()))

    def test_large_square_peg_does_not_fit(self):
        hole    = RoundHole(5.0)
        adapter = SquarePegAdapter(SquarePeg(8.0))
        self.assertFalse(hole.fits(adapter.radius()))

    def test_adapter_exact_boundary(self):
        # Square peg whose enclosing circle exactly equals the hole
        r = 5.0
        # width = r * 2 / sqrt(2) = r * sqrt(2)
        width   = r * math.sqrt(2)
        adapter = SquarePegAdapter(SquarePeg(width))
        self.assertAlmostEqual(r, adapter.radius(), places=6)

    # ------------------------------------------------------------------
    # SquarePeg
    # ------------------------------------------------------------------
    def test_square_peg_describe_contains_width(self):
        self.assertIn("6.5", SquarePeg(6.5).describe())


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_adapter.cpp
#include <gtest/gtest.h>
#include "pattern/adapter.hpp"
#include <cmath>

using namespace pattern;

static const double EPS = 1e-6;

// ---------------------------------------------------------------------------
// RoundHole / RoundPeg sanity
// ---------------------------------------------------------------------------
TEST(Adapter, RoundPegFitsInLargerHole)
{
    RoundHole hole(5.0);
    RoundPeg  peg(4.0);
    EXPECT_TRUE(hole.fits(peg.radius()));
}

TEST(Adapter, RoundPegDoesNotFitInSmallerHole)
{
    RoundHole hole(3.0);
    RoundPeg  peg(4.0);
    EXPECT_FALSE(hole.fits(peg.radius()));
}

TEST(Adapter, RoundPegFitsExactly)
{
    RoundHole hole(5.0);
    RoundPeg  peg(5.0);
    EXPECT_TRUE(hole.fits(peg.radius()));
}

// ---------------------------------------------------------------------------
// Object adapter: radius calculation
// ---------------------------------------------------------------------------
TEST(Adapter, AdapterRadiusIsHalfDiagonal)
{
    SquarePeg        peg(2.0);       // width = 2
    SquarePegAdapter adapter(peg);   // radius = 2 * sqrt(2) / 2 ≈ 1.414

    double expected = 2.0 * std::sqrt(2.0) / 2.0;
    EXPECT_NEAR(expected, adapter.radius(), EPS);
}

TEST(Adapter, SmallSquarePegFitsInBigHole)
{
    RoundHole        hole(5.0);
    SquarePeg        peg(4.0);       // enclosing circle radius ≈ 2.828
    SquarePegAdapter adapter(peg);
    EXPECT_TRUE(hole.fits(adapter.radius()));
}

TEST(Adapter, LargeSquarePegDoesNotFit)
{
    RoundHole        hole(5.0);
    SquarePeg        peg(8.0);       // enclosing circle radius ≈ 5.657
    SquarePegAdapter adapter(peg);
    EXPECT_FALSE(hole.fits(adapter.radius()));
}

// ---------------------------------------------------------------------------
// Class adapter
// ---------------------------------------------------------------------------
TEST(Adapter, ClassAdapterRadiusIsHalfDiagonal)
{
    SquarePegClassAdapter adapter(2.0);
    double expected = 2.0 * std::sqrt(2.0) / 2.0;
    EXPECT_NEAR(expected, adapter.radius(), EPS);
}

TEST(Adapter, ClassAdapterSmallPegFits)
{
    RoundHole             hole(10.0);
    SquarePegClassAdapter adapter(4.0);
    EXPECT_TRUE(hole.fits(adapter.radius()));
}

// ---------------------------------------------------------------------------
// SquarePeg describe
// ---------------------------------------------------------------------------
TEST(Adapter, SquarePegDescribeContainsWidth)
{
    SquarePeg peg(6.5);
    EXPECT_NE(std::string::npos, peg.describe().find("6.5"));
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
